runtime/domain.c

```diff
@@ -1533,22 +1544,16 @@ static void caml_poll_gc_work(void)
 /* FIXME: do not raise async exceptions */
 void caml_handle_gc_interrupt(void)
 {
-  atomic_uintnat* young_limit = domain_self->interruptor.interrupt_word;
   CAMLalloc_point_here;
 
   CAML_EV_BEGIN(EV_INTERRUPT_GC);
-  if (atomic_load_acq(young_limit) == INTERRUPT_MAGIC) {
-    /* interrupt */
+  if (caml_incoming_interrupts_queued()) {
     CAML_EV_BEGIN(EV_INTERRUPT_REMOTE);
-    while (atomic_load_acq(young_limit) == INTERRUPT_MAGIC) {
-      uintnat i = INTERRUPT_MAGIC;
-      atomic_compare_exchange_strong(
-          young_limit, &i, (uintnat)Caml_state->young_start);
-    }
     caml_handle_incoming_interrupts();
     CAML_EV_END(EV_INTERRUPT_REMOTE);
   }
 
+  caml_update_young_limit();
```

Comment:
I'm not sure about the logic here.

I think that what's going on is that we know we received an interrupt, we want to service it (except signals?), and "silence" the interrupt, but we also want to behave correctly if a new interrupt comes while we are servicing the first one. (At least we don't want to silence the signal for this second interrupt.)

I see broadly two approaches to do this:
1. We could silence the interrupt notification right away (by setting `young_limit` to `young_start`), before doing anything, and then service all interrupts received at this point. If we decide to not service an interrupt, we probably need to reset the notification (by setting `young_limit` again to a triggering value). If new interrupts come in the meantime, they will notify themselves to be handled next time.
2. Or we could service interrupts (without touching  young_limit`), and then before leaving silence the interrupt notification unless there are new interrupts queued.

Approach (1) feels more robust to me, but I think both could work in theory.

Before this commit, this function was clearly following approach (1): the INTERRUPT_MAGIC test would catch all service interrupts and the conditional would start by resetting `young_limit` and then handling the STW interrupts (only); and then the `caml_poll_gc_work` call would deal with other service interrupts and normal GC work. (Signal interrupts would get dropped, which may be unintended?)

After this commit, the code seems to do a weird mix of (1) and (2):
- first we check for STW interrupts and service them, without resetting `young_limit`, and only then do we call `caml_update_young_limit`; this sounds like approach (2) for STW interrupts.
- but after `caml_update_young_limit` we handle all other gc interrupts, this sounds like approach (1) for non-STW interrupts.

This sounds more complex than necessary. Am I missing something?

Reply:
From @gadmm explanation of `caml_update_young_limit`'s intended specification in https://github.com/ocaml/ocaml/pull/11095#discussion_r847276892, I suspect that there is a third approach which he may be considering: if each sort of interrupt is backed by an update to *some other runtime state* than the young limit, and `young_limit` is just an efficiency mechanism (it is meant to be equivalent to checking all the other parts in turn), then young_limit is not itself "the notification" as I had assumed earlier, and its value does not matter as long as it is correctly resynchronized with the rest of the runtime state once we are done dealing with interrupts. So there is no need for a global choice between approaches (1) and (2) hinted above, we can let each handler for a category of interrupt deal with its own runtime state, and then refresh the young limit the handlers are done.

Reply:
In fact, reading the whole diff, I can see that another call to update young_limit was added in this PR at the end of `caml_poll_gc_work`, suggesting that this third way is indeed the general approach.

Reply:
Yes, this is what I meant in the commit log of the 4th commit (young_limit does not carry information). I think this is how Jacques-Henri meant caml_update_young_limit.

Reply:
Ah, I understood your commit message as saying that there is no need to check against a specific INTERRUPT_MAGIC value, instead of any value above the next young pointer.